Lib/asyncio/selector_events.py

```diff
@@ -1061,16 +1063,58 @@ def write(self, data):
             self._loop._add_writer(self._sock_fd, self._write_ready)
 
         # Add it to the buffer.
-        self._buffer.extend(data)
+        self._buffer.append(data)
         self._maybe_pause_protocol()
 
-    def _write_ready(self):
+    def _write_sendmsg(self):
         assert self._buffer, 'Data should not be empty'
+        if self._conn_lost:
+            return
+        try:
+            n = self._sock.sendmsg(self._buffer)
+            self._adjust_leftover_buffer(n)
+        except (BlockingIOError, InterruptedError):
+            pass
+        except (SystemExit, KeyboardInterrupt):
+            raise
+        except BaseException as exc:
+            self._loop._remove_writer(self._sock_fd)
+            self._buffer.clear()
+            self._fatal_error(exc, 'Fatal write error on socket transport')
+            if self._empty_waiter is not None:
+                self._empty_waiter.set_exception(exc)
+        else:
+            self._maybe_resume_protocol()  # May append to buffer.
+            if not self._buffer:
+                self._loop._remove_writer(self._sock_fd)
+                if self._empty_waiter is not None:
+                    self._empty_waiter.set_result(None)
+                if self._closing:
+                    self._call_connection_lost(None)
+                elif self._eof:
+                    self._sock.shutdown(socket.SHUT_WR)
 
+    def _adjust_leftover_buffer(self, n: int, /) -> None:
+        buffer = self._buffer
+        while n:
+            b = buffer.popleft()
+            b_len = len(b)
+            if b_len <= n:
+                n -= b_len
+            else:
+                buffer.appendleft(b[n:])
+                break
+
+    def _write_send(self):
+        assert self._buffer, 'Data should not be empty'
         if self._conn_lost:
             return
         try:
-            n = self._sock.send(self._buffer)
+            buffer = self._buffer.popleft()
+            n = self._sock.send(buffer)
```

Comment:
An ability to send the merged buffer is lost here.
The change was visible when I had experimented with aiohttp performance optimization.
Ideally, we should use `deque` if `sendmsg()` is available and `bytearray` if `send()` is used.
It might require two separate `_SelectorSocketTransport` classes, especially if we can select the proper implementation by the static flag check.

Reply:
So should I create a separate Transport and based on sendmsg presence use it ?

Reply:
But if bytearray is used then it would cause memory copy and that can be avoided if memoryview and deque is used.

Reply:
This is a tradeoff between the memory copy and multiple syscall calls.
In my experience, the first is better. Also, the very first asyncio implementation used deque. Then it was replaced with bytearray for the sake of performance.
Sure, `sendmsg()` is event better. I'm talking about the fallback now.

Reply:
I benchmarked it and deque with memoryview is faster than bytearray, bytearray is slow because of overhead of allocating it and copying the data and resizing the bytearray.
Script:
``` py
import asyncio
from timeit import timeit
async def handle_echo(reader: asyncio.StreamReader, writer: asyncio.StreamWriter):
    data = b'o' * 1024 * 1024 * 100
    writer.write(data)
    await writer.drain()
    writer.close()

async def main():
    server = await asyncio.start_server(
        handle_echo, '127.0.0.1', 8882)

    async with server:
        asyncio.create_task(server.start_serving())
        reader, writer = await asyncio.open_connection('127.0.0.1', 8882)
        while True:
            data = await reader.read(1024 * 1024)
            if not data:
                break
        writer.close()
        await writer.wait_closed()

print(timeit(lambda: asyncio.run(main()), number=10))
```
Set `HAVE_SENDMSG` to `False` and you can try it.

Reply:
Your benchmark is not complete. 
Instead of sending the huge buffer at once
```
    data = b'o' * 1024 * 1024 * 100
    writer.write(data)
    await writer.drain()
```
try sending many smaller chunks
```
    data = b'o' * 1024  * 100
    for num in range(1024):
        writer.write(data)
        await writer.drain()
```
It is a much more realistic scenario.
A typical code doesn't send gigabytes of data at a single call but uses multiple chunked messages. Any web protocol I worked with supports this mode.

The discussion for queue -> bytearray is lost; at least I cannot find it. GitHub has the only commit itself without an external link to bpo (or google code where tulip was initially developed).

Reply:
I mean this commit: https://github.com/python/cpython/commit/a5062c5d81ba4d955edba55ff74d63110f1004a4

Reply:
I think we should make it configurable with a `use_vectored_io` argument which on supported platforms would use sendmsg otherwise fallback to regular `send`. What do you think? @asvetlov

Reply:
Also as a datapoint uvloop always uses a queue to avoid data copying and uses sendmsg.

See https://github.com/MagicStack/uvloop/blob/c808a663b297bb2aee745523c277d6fafecebbeb/uvloop/handles/stream.pyx#L421

See https://github.com/libuv/libuv/blob/03f1a6979caff380bfcfcd1293390474eaad6164/src/unix/stream.c#L868

Reply:
Correct. 
`uvloop` uses `libuv`.
`libuv` provides multi-buffer API only.
For this mode, a queue is optimal if `sendmsg()` is available.

I'm speaking about the fallback mode with only `send()` can be used.
In such fallback, `bytearray` is more performant than `queue`, that's why asyncio uses it right now.

I propose the following:
1. Detect `sendmsg()` availability at the import time
2. If `sendmsg()` is present, use `queue` and `sendmsg`.
3. If fallback is required, use `bytearray` and `send`.
4. For source code simplification, (2) and (3) could be implemented as two different private classes (maybe with a common private predecessor). The desired implementation could be selected by `loop._make_socket_transport` / `loop._make_ssl_transport`

SSL seems never use bulk writes, it could be discovered and solved by a separate pull requests.

Reply:
SGTM, I'll implement it by this weekend.

Reply:
It looks like the code in this PR still writes one buffer at a time. I agree with Andrew that if we're using `_write_send` it is better to copy the data (maybe in `write`?) than to incur the overhead of potentially many syscalls. Or did I miss something?

Reply:
done it now joins the data in a bytearray and then writes with `send`

Reply:
The syscall overhead is not as much, send is still faster than it was before this because of less memory copying here but less than sendmsg as expected.
https://github.com/python/cpython/issues/91166#issuecomment-1288541370

Reply:
I'll think more about this though.

Reply:
I worry about cases where people write many small strings. Extreme case:
```py
for c in b'x'*1000000:
    t.write(c)
```
It would be nice to know at what point the syscall overhead overwhelms the gains, on platforms without `sendmsg` (which includes Mac and Windows)?

Reply:
I'll run some benchmarks for this.

Reply:
Okay, waiting for those benchmarks.

Reply:
Benchmark:
```py

import asyncio
from pyperf import Runner


async def handle_echo(reader: asyncio.StreamReader, writer: asyncio.StreamWriter):
    for _ in range(10000):
        writer.write(b'x')
        await writer.drain()
    writer.close()
    await writer.wait_closed()


async def main():
    server = await asyncio.start_server(
        lambda reader, writer: handle_echo(reader, writer), '127.0.0.1', 8882)

    async with server:
        asyncio.create_task(server.start_serving())
        reader, writer = await asyncio.open_connection('127.0.0.1', 8882)
        while True:
            data = await reader.read(1)
            if not data:
                break
        writer.close()
        await writer.wait_closed()

if __name__ == '__main__':
    runner = Runner()
    runner.bench_async_func('benchmark', main)
```

Result:
```
Benchmark hidden because not significant (1): benchmark
```

This is as expected because there is fast path for this kind of input. See https://github.com/python/cpython/blob/2c62bcb4bc9449937ad46288665ba36457beb681/Lib/asyncio/selector_events.py#L1071